src/Bicep.Core/TypeSystem/Az/AzResourceTypeLoaderFactory.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Linq;
+using Azure.Bicep.Types.Az;
+using Bicep.Core.Syntax;
+using System.IO;
+using Bicep.Core.Registry.Oci;
+using Newtonsoft.Json;
+using Bicep.Core.Features;
+
+namespace Bicep.Core.TypeSystem.Az
+{
+    public class AzResourceTypeLoaderFactory : IAzResourceTypeLoaderFactory
+    {
+        private readonly IFeatureProviderFactory featureProviderFactory;
+
+        private Dictionary<string, IAzResourceTypeLoader> resourceTypeLoaders = new(){
+            {"builtin", new AzResourceTypeLoader(new AzTypeLoader())},
+        };
+
+        public AzResourceTypeLoaderFactory(IFeatureProviderFactory featureProviderFactory)
+        {
+            this.featureProviderFactory = featureProviderFactory;
+        }
+
+        public IAzResourceTypeLoader? GetResourceTypeLoader(ImportDeclarationSyntax? ids, IFeatureProvider features)
+        {
+            if (!features.DynamicTypeLoading || ids is null)
+            {
+                return resourceTypeLoaders["builtin"];
+            }
+
+            var azProviderDir = Path.Combine(features.CacheRootDirectory, "br", "mcr.microsoft.com", @"bicep$providers$az", ids.Specification.Version);
+            var ociManifestPath = Path.Combine(azProviderDir, "manifest.json");
+            if (!File.Exists(ociManifestPath))
+            {
+                return null;
+            }
+
+            // Read the OCI manifest
+            var manifestFileContents = File.ReadAllText(ociManifestPath);
+            OciManifest? ociManifest = JsonConvert.DeserializeObject<OciManifest>(manifestFileContents);
+
+            if (ociManifest is null)
+            {
+                return null;
+            }
+
+            // Get the filename of the OCI type definitions artifact
+            var typesDefinitionFilename = ociManifest.Layers.SingleOrDefault()?.Annotations["org.opencontainers.image.title"];
```

Comment:
>org.opencontainers.image.title" [](http://example.com/codeflow?start=93&length=31)

This annotation is intended to be a human-readable description. We should probably find a standard OCI one that fits better or create a custom one. 

Personally, we should enforce a specific naming convention (hard-coded file name) and not allow deviations from it until we have a good use case for it. Otherwise, we're just adding complexity to both publish and restore process that may not be needed.

Reply:
Lets talk about it a bit more, I will set up some time

Reply:
Sounds good.

Reply:
I suspect we can close on that over Teams and avoid a meeting.

Reply:
I am using a constant now in the file, I may end upo with a similar pattern like the one in the modules but for now I think it should work